storage/series/series.go

```diff
@@ -452,14 +468,68 @@ func (s *dbSeries) OnRetrieveBlock(
 		Checksum: digest.SegmentChecksum(segment),
 	}
 	b.ResetRetrievable(startTime, s.blockRetriever, metadata)
-	b.OnRetrieveBlock(id, startTime, segment)
+	// Use s.id instead of id here, because id is finalized by the context whereas
+	// we rely on the G.C to reclaim s.id. This is important because the block will
+	// hold onto the id ref, and (if the LRU caching policy is enabled) the shard
+	// will need it later when the WiredList calls its OnEvictedFromWiredList method.
+	// Also note that ResetRetrievable will mark the block as not retrieved from disk,
+	// but OnRetrieveBlock will then properly mark it as retrieved from disk so subsequent
+	// calls to WasRetrievedFromDisk will return true.
+	b.OnRetrieveBlock(s.id, startTime, segment)
 
 	// NB(r): Blocks retrieved have been triggered by a read, so set the last
 	// read time as now so caching policies are followed.
 	b.SetLastReadTime(s.now())
 
 	// If we retrieved this from disk then we directly emplace it
-	s.blocks.AddBlock(b)
+	s.addBlock(b)
+
+	if list := s.opts.DatabaseBlockOptions().WiredList(); list != nil {
+		// Need to update the WiredList so blocks that were read from disk
+		// can enter the list (OnReadBlock is only called for blocks that
+		// were read from memory, regardless of whether the data originated
+		// from disk or a buffer rotation.)
+		list.Update(b)
+	}
+}
+
+// OnReadBlock is only called
```

Comment:
Comment seems not to be 100% complete here?

Reply:
yeah looks like I included it as an inline comment above, fixed it.